### PR TITLE
Filter low gas txns early

### LINKED ISSUE #4793: TxPool further improvements
It seems that we made quite significant optimization here: https://github.com/NethermindEth/nethermind/pull/4702

I added metric: PendingTransactionsWithExpensiveFiltering. Ideally, we should filter as many transactions as possible before we reach this metric. 

One potential idea:
When our TxPool is full we shouldn't accept Tx which are worse (in terms of GasPrice) than our worse transaction in TxPool. We're doing something like this in FeeTooLowTxFilter: https://github.com/NethermindEth/nethermind/blob/master/src/Nethermind/Nethermind.TxPool/TxPool.cs#L119

However, it should be implemented without accessing the state. If we're accessing state we won't have benefits of optimization.

### PR DESCRIPTION
**Pull request by benaadams**
Fixes Closes Resolves #4793

## Changes

* Changes `FeeTooLowFilter` for `BalanceTooLowFilter`
* Introduces `FeeTooLowFilter` early which just considers gas fee and not balance

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a fix or a feature that would cause existing functionality not to work as expected)
- [ ] Documentation update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional or API changes)
- [ ] Build-related changes
- [ ] Other: _description_ 

#### Requires testing

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by marcindsobczak — APPROVED** [review]
LGTM!

I checked grafana and made some calculations.
We are discarding ~76% of all received txs. Most of them in `AlreadyKnownFilter` (~58% of all received).
In `FeeTooLow` we discard ~13% of received. This filter is after `AlreadyKnown` so most of txs are discarded earlier. In this filter we are discarding ~30% of remaining txs. My intuition says that most of them are discarded because of low price and only minority because of balance and we can reduce state accesses by even 25-30%, that's great! And after adding new metric (more in comment) we will not need to guess. But it can be done in a separate PR

File: src/Nethermind/Nethermind.TxPool/Filters/BalanceTooLowFilter.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            IReleaseSpec spec = _specProvider.GetCurrentHeadSpec();
+            UInt256 balance = state.SenderAccount.Balance;
+            UInt256 affordableGasPrice = tx.CalculateAffordableGasPrice(spec.IsEip1559Enabled, _headInfo.CurrentBaseFee, balance);
+
+            if (_txs.TryGetLast(out Transaction? lastTx)
+                && affordableGasPrice <= lastTx?.GasBottleneck)
+            {
+                Metrics.PendingTransactionsTooLowFee++;
+                if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, too low payable gas price with options {handlingOptions} from {new StackTrace()}");
+                return AcceptTxResult.FeeTooLow.WithMessage($"FeePerGas needs to be higher than {lastTx.GasBottleneck.Value} to be added to the TxPool. Affordable FeePerGas of rejected tx: {affordableGasPrice}.");
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/Filters/BalanceTooLowFilter.cs]: Added `InsufficientFundsToCompete`

Existing metric
`PendingTransactionsTooLowFee`
Also two new metrics
`PendingTransactionsBalanceToLowToCompeteOnFee`
`PendingTransactionsPassedFiltersButCannotCompeteOnFees`

**Comment by benaadams** [comment]
Keccak is kinda expensive; so a follow up could be to delay calculating the Txn hash

Then rearrange the filters from
```cs
_filterPipeline.Add(new NullHashTxFilter());
_filterPipeline.Add(new AlreadyKnownTxFilter(_hashCache, _logger));
_filterPipeline.Add(new FeeTooLowFilter(_headInfo, _transactions, logManager));
_filterPipeline.Add(new MalformedTxFilter(_specProvider, validator, _logger));
_filterPipeline.Add(new GasLimitTxFilter(_headInfo, txPoolConfig, _logger));
_filterPipeline.Add(new UnknownSenderFilter(ecdsa, _logger));
```
to
```cs
_filterPipeline.Add(new FeeTooLowFilter(_headInfo, _transactions, logManager));
_filterPipeline.Add(new MalformedTxFilter(_specProvider, validator, _logger));
_filterPipeline.Add(new NullHashTxFilter()); // change to calculate the hash
_filterPipeline.Add(new AlreadyKnownTxFilter(_hashCache, _logger));
_filterPipeline.Add(new GasLimitTxFilter(_headInfo, txPoolConfig, _logger));
_filterPipeline.Add(new UnknownSenderFilter(ecdsa, _logger));
```
Depending if rejecting the duplicates is cheaper than not calculating the hash at first and rejecting on fees and malformed; prior to checking for duplicates (and calculating the hash)

**Comment by benaadams** [comment]
> Depending if rejecting the duplicates is cheaper than not calculating the hash at first and rejecting on fees and malformed; prior to checking for duplicates (and calculating the hash)

However the deserialise calculates it straight on the byte stream and later has to build the stream from the txn, so isn't a straightforward a change (as well as plumbing through txns without hashes)

**Comment by benaadams** [comment]
> > Depending if rejecting the duplicates is cheaper than not calculating the hash at first and rejecting on fees and malformed; prior to checking for duplicates (and calculating the hash)
> 
> However the deserialise calculates it straight on the byte stream and later has to build the stream from the txn, so isn't a straightforward a change (as well as plumbing through txns without hashes)

@marcindsobczak I experimented with late hashing in https://github.com/NethermindEth/nethermind/pull/5176 if you want to see what the rate of rejection stage %s look like?

**Comment by benaadams** [comment]
Actually https://github.com/NethermindEth/nethermind/pull/5176 seems ok; if the metrics hold up

**Comment by benaadams** [comment]
> Actually #5176 seems ok; if the metrics hold up

Metrics don't seem to be worth it with the added complexity, have closed that issue

**Comment by tkstanczak** [comment]
Was it not designed to discard transactions without hash?
Is it worth to check anything on malformed transactions? Will it not cause crashes?

**Comment by benaadams** [comment]
> Was it not designed to discard transactions without hash? Is it worth to check anything on malformed transactions? Will it not cause crashes?

Txns without hashes are discarded; however its Nethermind that calculates the hash. I experimented with delay calculating the hash in https://github.com/NethermindEth/nethermind/pull/5176 however didn't have good evidence it helped

Malformed txns are checked

**Pull request closed without merging**